### PR TITLE
Kaby Lake modesetting support

### PR DESCRIPTION
**Pull request by no92**
This PR is a massive dump of code to get modesetting working on Kaby Lake. eDP and HDMI output are supported, and support for regular DisplayPort should work but is untested so far.

For now, the default is to limit the outputs to a single one (the first one that is successfully pre-enabled). This should improve in a future PR in order to avoid blowing this PR up even more and to get this upstreamed after ~2 years of work.

Most code that is used for Kaby Lake support has been converted to C++, and a few utilities have been added to aid with debugging.

### CODE REVIEW

File: src/kaby_lake/ddi-translations.cpp [hunk 1]
```diff
@@ -0,0 +1,162 @@
+#include "src/kaby_lake/ddi-translations.hpp"
+#include "src/helpers.h"
+
+/* Kabylake H and S */
+static const struct lil_ddi_buf_trans_entry _kbl_trans_dp[] = {
```
Comment (ElectrodeYT) [inline src/kaby_lake/ddi-translations.cpp:5]: Can we maybe note the source of these structs? Would be helpful if/when we add more for other gens

File: src/kaby_lake/plane.cpp [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+}
+
+}
+
+void enable(LilGpu *gpu, LilCrtc *crtc, bool vblank_wait) {
+	uint32_t htotal = crtc->current_mode.htotal;
+	uint32_t pixel_clock = crtc->current_mode.clock;
+
+	uint8_t wm[8];
+	wm[0] = (gpu->mem_latency_first_set >> 0) & 0xFF;
```
Comment (ElectrodeYT) [inline src/kaby_lake/plane.cpp:50]: The compiler should optimize these away, but the `& 0xFF` _should_ not be needed here, as far as i know, as the array that is written to is a `uint8_t`

File: src/kaby_lake/pll.cpp [hunk 3]
```diff
@@ -0,0 +1,361 @@
...
+#include "src/kaby_lake/hdmi.hpp"
+#include "src/kaby_lake/pcode.hpp"
+#include "src/kaby_lake/pll.hpp"
+#include "src/regs.h"
+
+namespace kbl::pll {
+
+// TODO(CLEAN;BIT): this function needs to be cleaned up
+// 					specifically, we should be using enums or defines for this bit setting/clearing
+LilError disable(LilGpu *gpu, LilConnector *con) {
```
Comment (ElectrodeYT) [inline src/kaby_lake/pll.cpp:14]: This function has unclear bit manipulation going on

File: src/kaby_lake/pll.cpp [hunk 4]
```diff
@@ -0,0 +1,361 @@
...
+	uint32_t link_rate = (REG(DPLL_CTRL1) & mask) >> shift;
+
+	if(con->type == EDP && (link_rate == 4 || link_rate == 5)) {
+		lil_panic("eDP PLL disable logic unimplemented");
+	}
+
+	return LilError::LIL_SUCCESS;
+}
+
+void dpll_clock_set(LilGpu *gpu, LilCrtc *crtc) {
```
Comment (ElectrodeYT) [inline src/kaby_lake/pll.cpp:54]: This function has unclear bit manipulation going on

File: src/kaby_lake/pll.cpp [hunk 5]
```diff
@@ -0,0 +1,361 @@
...
+		default:
+			lil_panic("invalid PLL");
+	}
+
+	return (REG(DPLL_STATUS) & mask) != 0;
+}
+
+} // namespace
+
+void dpll_ctrl_enable(LilGpu *gpu, LilCrtc *crtc, uint32_t link_rate) {
```
Comment (ElectrodeYT) [inline src/kaby_lake/pll.cpp:201]: This function has unclear bit manipulation going on

**Review by ElectrodeYT — COMMENTED** [review]
Quick look, still need to look a bit closer at this, but have come up with a few nitpicks

File: src/kaby_lake/plane.cpp (hunk 2 above)
Comment (d-tatianin) [reply src/kaby_lake/plane.cpp:50]: yes but it helps readability IMO

Comment (ElectrodeYT) [reply src/kaby_lake/plane.cpp:50]: understandable

**Review by ElectrodeYT — APPROVED** [review]

**Review by Dennisbonke — APPROVED** [review]
LGTM

**Pull request merged**